lib/edge-runtime/missingParamCorrection.ts

```diff
@@ -0,0 +1,79 @@
+import { OpenAPIV3_1 } from "openapi-types";
+import { FunctionCall } from "@superflows/chat-ui-react";
+import { Action } from "../types";
+import requestCorrectionPrompt from "../prompts/requestCorrection";
+import { bodyPropertiesFromRequestBodyContents } from "./requests";
+
+import { getOpenAIResponse } from "../queryOpenAI";
+import { ChatGPTMessage } from "../models";
+
+export async function getCorrectionsForMissingCommandArgs(
+  action: Action,
+  command: FunctionCall,
+  previousConversation: ChatGPTMessage[]
+): Promise<{
+  corrections: { [param: string]: "ask user" | any };
+  newSystemMessages: ChatGPTMessage[] | null;
+}> {
+  // TODO: name this function better
+
+  let bodyRequired: string[] = [];
+
+  if (action.request_body_contents) {
+    const { schema } = bodyPropertiesFromRequestBodyContents(
+      action.request_body_contents
+    );
+    bodyRequired = schema?.required || [];
+  }
+
+  const queryRequired = getRequiredQueryParams(action);
+  const allRequiredParams = bodyRequired.concat(queryRequired);
+
+  const missingParams = allRequiredParams.filter(
+    (param) => !command.args[param]
+  );
+
+  let correctionPrompt: ChatGPTMessage[] | null = null;
+  const corrections: { [param: string]: "ask user" | string } = {};
+  for (const param of missingParams) {
+    const missingParamRes = await getMissingParam(
+      param,
+      action,
+      previousConversation
+    );
+    if (missingParamRes.response) corrections[param] = missingParamRes.response;
+    correctionPrompt = missingParamRes.correctionPrompt;
+  }
+  return { corrections, newSystemMessages: correctionPrompt };
+}
+
+async function getMissingParam(
+  missingParam: string,
+  action: Action,
+  previousConversation: ChatGPTMessage[]
+): Promise<{
+  response: string | null;
+  correctionPrompt: ChatGPTMessage[] | null;
+}> {
+  console.log(`Parameter ${missingParam} is missing attempt to get it`);
+  const correctionPrompt = requestCorrectionPrompt(missingParam, action);
+  if (!correctionPrompt) return { response: null, correctionPrompt: null };
+  const prompt = [...previousConversation].concat(correctionPrompt);
+  console.log("Request correction prompt:\n", prompt);
+  let response = await getOpenAIResponse(prompt, undefined, "3");
```

Comment:
Also we should think carefully about making sure we don't hit the context limit here - standard 3.5 has half the limit that 4 does (another reason to reduce the max output tokens, since this uses up space for input tokens).

I think it's probably safe to cut long function messages to the 4K token limit here (although have to be careful this won't cut the same messages for the object in the `Angela` function!)

Reply:
this look good?